doc/sphinx/user-extensions/syntax-extensions.rst

```diff
@@ -981,106 +1003,98 @@ notations are given below. The optional :n:`@scope` is described in
 
-.. cmd:: Local Declare Custom Entry @ident
+   .. todo PR dropped "at the OCaml level" above--not needed, right?
 
-   Custom entries survive modules unless the command ``Local Declare
-   Custom Entry`` is used instead of :cmd:`Declare Custom Entry`.
+   .. insertprodn scope scope_key
 
-.. _Scopes:
+   .. prodn::
+      scope ::= @scope_name
+      | @scope_key
+      scope_name ::= @ident
+      scope_key ::= @ident
 
-Interpretation scopes
-----------------------
+Each notation scope has a single :token:`scope_name`, which by convention
+ends with the suffix "_scope", as in "nat_scope".  One or more :token:`scope_key`\s
+(delimiting keys) may be associated with a notation scope with the :cmd:`Delimit Scope` command.
+Most commands use :token:`scope_name`; :token:`scope_key`\s are used within :token:`term`\s.
 
-   .. insertprodn scope scope
+.. cmd:: Declare Scope @scope_name
 
-   .. prodn::
-      scope ::= @ident
-
-An *interpretation scope* is a set of notations for terms with their
-interpretations. Interpretation scopes provide a weak, purely
-syntactical form of notation overloading: the same notation, for
-instance the infix symbol ``+``, can be used to denote distinct
-definitions of the additive operator. Depending on which interpretation
-scopes are currently open, the interpretation is different.
-Interpretation scopes can include an interpretation for numerals and
-strings, either at the OCaml level or using :cmd:`Numeral Notation`
-or :cmd:`String Notation`.
-
-.. cmd:: Declare Scope @scope
-
-   This adds a new scope named :n:`@scope`. Note that the initial
-   state of Coq declares by default the following interpretation scopes:
+   Declares a new notation scope. Note that the initial
+   state of Coq declares the following notation scopes:
    ``core_scope``, ``type_scope``, ``function_scope``, ``nat_scope``,
    ``bool_scope``, ``list_scope``, ``int_scope``, ``uint_scope``.
 
-The syntax to associate a notation to a scope is given
-:ref:`above <NotationSyntax>`. Here is a typical example which declares the
-notation for conjunction in the scope ``type_scope``.
-
-.. coqtop:: in
-
-   Notation "A /\ B" := (and A B) : type_scope.
-
-.. note:: A notation not defined in a scope is called a *lonely*
-          notation. No example of lonely notations can be found in the
-          initial state of Coq though.
-
+   Use commands such as :cmd:`Notation` to add notations to the scope.
 
 Global interpretation rules for notations
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. todo the evidence suggests that scopes are not organized as a stack,
+   so it's incorrect to refer to a stack of scopes below
+   1. One would expect the follow sequence would give an error because the last
+   command is not a pop.
+     Require Import ZArith.
+     Require Import PrimFloat.
+     Open Scope Z_scope.
+     Open Scope float_scope.
+     Close Scope Z_scope.
+   2. Print Visibility shows that items in the middle of the stack can be removed.
```

Comment:
Indeed, `Close Scope` operates in the middle of the "scope stack".  But it's impossible to insert a new scope in the middle (you have to re-open all the ones after it again), and printing generally prefers the most recently opened stack, and parsing always does.

Reply:
Explained the "stack"  more in the text and dropped the todo.